### PR TITLE
Implement 'rewrite' command to exclude files from existing snapshots

### LINKED ISSUE #14: Delete Files from Existing Snapshot
In cases of accidential backup of e.g. too large files, I would like to be able to delete specific files or directories (incl. recursion) from existing snapshots

### PR DESCRIPTION
**Pull request by dionorgua**
What is the purpose of this change? What does it change?
--------------------------------------------------------

This adds 'rewrite' command that provides a way to remove data from existing snapshots.

This is some sort of preliminary pull requests. It's actually works here on test repo, but lacks tests, etc. Mostly to discuss things

Was the change discussed in an issue or in the forum before?
------------------------------------------------------------

[ replaces #2720 because it was created from master branch ]
closes #14 

Checklist
---------

- [x] I have read the [Contribution Guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches)
- [x] I have enabled [maintainer edits for this PR](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork)
- [ ] I have added tests for all changes in this PR
- [ ] I have added documentation for the changes (in the manual)
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (template [here](https://github.com/restic/restic/blob/master/changelog/TEMPLATE))
- [x] I have run `gofmt` on the code in all commits
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits)
- [ ] I'm done, this Pull Request is ready for review

### CODE REVIEW

File: cmd/restic/cmd_rewrite.go
Comment (aawsome) [inline cmd/restic/cmd_rewrite.go:203]: The index is now automatically saved by `repo.Flush()` #2773, no need to call `SaveIndex` after `Flush` any more.

File: cmd/restic/cmd_rewrite.go
Comment (aawsome) [inline cmd/restic/cmd_rewrite.go:157]: You could also use the TreeSaver from `internal/archiver` to make saving parallel. Can be however postponed to a later optimization.

File: cmd/restic/cmd_rewrite.go
Comment (aawsome) [inline cmd/restic/cmd_rewrite.go:196]: You only need to flush the repo when the snapshot was modified.

I also wouldn't flush after finishing a single snapshot, but once after all snapshots are rewritten, i.e. in line 289.
The reason is that flush may finish small packs. If you modify a lot of snapshots generating only few new tree blobs, you may end up in having lots of new packs which each only contain few tree blobs.

However, the snapshot files should only be written after flush finishes with nil error. So also writing the snapshots should be postponed. I suggest returning the new snapshot object here and write all snapshots after `Flush` in `runRewrite`

File: cmd/restic/cmd_rewrite.go
Comment (aawsome) [inline cmd/restic/cmd_rewrite.go:225]: I would suggest to remove the deletion code totally from `cmd_rewrite.go` and instead call `runForget` from `forget.go` with the snapshot-IDs to remove.

That way you will not have any critical code that removes anything here, but keep that "burden" in exclusive pieces of code (which is well-tested and can be optimized or maybe turned into 2-phase-compaction later more easily).

Also allows to add a `--prune` flag here and pass this directly to `runForget`; we should then also add the `addPruneFlags()` call once #2718 is merged.

File: cmd/restic/cmd_rewrite.go
Comment (aawsome) [inline cmd/restic/cmd_rewrite.go:258]: If you outsource the deletion of files, a non-exclusive lock would suffice here.

**Review by aawsome — COMMENTED** [review]
Looks mostly good to me :+1: 
I propose to call the existing code from `cmd_forget` to actually remove snapshots. That way also a pruning could be directly started from this command which IMO improves the usage for end users.

**Comment by dionorgua** [comment]
Thanks a lot for review. I'll try to address everything soon

**Comment by edmwagner** [comment]
I Just mirrored my minio and used this branch to rewrite a few different scenarios.  looks good so far.  restoring files worked as expected and matched versions restored from original repo.  :)  The repo is several GB smaller though.



I would like to help move this feature forward.

**Comment by dionorgua** [comment]
Hi,
I was busy with personal things. Will try to catch up this weekend and at least update/rebase it.

**Comment by Komic** [comment]
Is supporting such a feature for a whole repository planned?

**Comment by aawsome** [comment]
> Is supporting such a feature for a whole repository planned?

This PR already allows to give multiple snapshots or even multiple snapshot criteria. E.g. you can run it with all snapshots or give multiply `-H` options to cover all hosts.

We could add a possibility to specify `all` which resolves to all snapshots - but I would leave that to a future PR as it is pretty independent from this change.

**Comment by Komic** [comment]
Good stuff, thank you. 👍

**Comment by bertbesser** [comment]
What's the status of this PR? Any chance that it gets merge soon?

**Comment by rawtaz** [comment]
@bertbesser This PR should be considered work in progress. No decision has been made about whether this should be merged or not, the PR is not yet marked as done by the author, and there's a bunch of unaddressed comments. Clearly it's not ready for merging, and the status is simply what you see here in the GitHub GUI where you posted a comment. That said, there's a lot of other stuff going on if you look in this repository, so consider this WIP. Feel free to try it yourself though, just don't consider it production ready (at least not officially).

**Comment by dionorgua** [comment]
Hi,

@bertbesser 

It was working pretty well on my machine in time it was created (but yes, it's very experimental). Good news is that it doesn't touch any data pack files at all. Basically It just adds new snapshot as copy of existing with a few files removed (plus optionally removes old snapshot). I don't think that it can corrupt repository (except new snapshots, that can be just deleted manually)

I think that 'rewrite + check + prune' scenario should be safe enough. I've already used it on my own 2.5TB repository without any problems. Removed most of garbage from old snapshots.

But I'm not sure that it'll work as is right now, I'll rebase it on top of latest git and try to address most of comments next week.

**Comment by edmwagner** [comment]
Absolutely!  I’ve also used it to clean up 30gb of junk from early
backups.  I didn’t know how to filter back then.  My repo was 40 gb and now
rest consistently under 10 with about 2years of backups.  Really helped!!!
But that was also some time ago.

**Comment by komic** [comment]
Hi @dionorgua, is there any chance you'd have an update regarding your work on this PR? I may or may have not oopsied a B2 backup again. 🤠

**Comment by MichaelEischer** [comment]
I've rebased the PR and made quite a few cleanups (see the commit messages for more details). The main tasks left are:
- [ ] Writing some documentation
- [ ] More tests. The current test coverage is just the absolute minimum
- [ ] Decide what to do with the `--inplace` flag, I'll probably rename it to `--forget`
- [ ] Performance optimizations.
- [ ] Finalize the CLI options

File: cmd/restic/cmd_rewrite.go
Comment (MichaelEischer) [reply cmd/restic/cmd_rewrite.go:67]: Adding all prune options would clutter the options list for the `rewrite` command. I think it's better to keep them separate.

File: cmd/restic/cmd_rewrite.go
Comment (MichaelEischer) [reply cmd/restic/cmd_rewrite.go]: That would also require the BlobSaver. It's probably better to integrate some form of asynchronous blob submission directly into the repository and then use that.

File: cmd/restic/cmd_rewrite.go
Comment (MichaelEischer) [reply cmd/restic/cmd_rewrite.go:139]: Chaining the commands is probably more complex than directly removing the snapshot. In particular the error handling, log output etc. would have to be coordinated between the commands.

File: cmd/restic/cmd_rewrite.go
Comment (MichaelEischer) [reply cmd/restic/cmd_rewrite.go]: Upgrading a lock in the middle of a command is a problem. There is no guarantee that this will work.

**Comment by NovacomExperts** [comment]
I'm so thrilled that this PR is getting some progress ! Many thanks to the team members for all the latest commits. I will do some test on my mostly "local" repos.

I support @MichaelEischer on the naming for `--forget` instead of `--inplace`. Of course, "are you sure" could follow before execution, just like prune. As discussed with @dionorgua before, a clear `--tag-old-snapshots-with <tag>` and `--tag-new-snapshots-with <tag>` would work great for managing 2 step forget process. 

Keep on the great work !

**Comment by bertbesser** [comment]
> Hi,
> 
> @bertbesser
> 
> It was working pretty well on my machine in time it was created (but yes, it's very experimental). Good news is that it doesn't touch any data pack files at all. Basically It just adds new snapshot as copy of existing with a few files removed (plus optionally removes old snapshot). I don't think that it can corrupt repository (except new snapshots, that can be just deleted manually)
> 
> I think that 'rewrite + check + prune' scenario should be safe enough. I've already used it on my own 2.5TB repository without any problems. Removed most of garbage from old snapshots.
> 
> But I'm not sure that it'll work as is right now, I'll rebase it on top of latest git and try to address most of comments next week.

@NovacomExperts The `--tag-xxx-snapshots-with` would just be a welcome addition to this, correct? Not a fundamentally other approch if I get you right. Cheers

**Comment by MichaelEischer** [comment]
I've added enough tests to achieve over 80% test coverage for the new code. There's now also a new documentation section. I'll probably split some of the preparatory refactoring commits into separate PRs to reduce the amount of code a bit.

> As discussed with @dionorgua before, a clear `--tag-old-snapshots-with <tag>` and

`--tag-old-snapshots` is impossible to implement. We cannot change the tag of an existing snapshot without  creating a new one and removing the old snapshot.

Instead of using tags, using the printed snapshot id should also work.

**Comment by MichaelEischer** [comment]
I've added a commit to let the `rewrite` command reject rewriting anything which would result in loosing data in a tree. This ensures that we can add new fields to the `Tree` struct without having to increase the repository version every time. Without the check, the `rewrite` command could just loose data.

The check is not completely optimal, but it should prevent rewriting for all problematic situations. If the need arises to handle this in a more targeted way, we can still adapt the implementation later on.

**Comment by NovacomExperts** [comment]
@MichaelEischer This is such a great addition to Restic. Again, thank you a thousand times for making this possible. At a point I was almost ready to learn GO and give it a shot, but seeing you commits, and the level of expertise and care needed... Man this is waaaay out my league :)

`--tag-old-snapshots` was merely a brainstorming idea. I think that the --forget  flag will do the job perfectly in one time.

I'll have some free time to in the next few days to recompile with the latest commits. I'll test a few of my "large" repos.

**Comment by NovacomExperts** [comment]
Hello !

I've been playing around with [72196e5](https://github.com/restic/restic/commit/72196e54c0e71d4d2ae26314173f57914b09ba28) and I had to use in production on a version 1 repo. It was a 3 host local repo composed of 830 GB and was correctly shrinked to 319 GB. Everything went very well and all checks are ok (including read-data). Restore tests were OK too. I have kept a backup of the original repo just in case but so far so good ! I wanted to exclude a specific folder from multiple snapshots **except the last 30 days** on a specific host. Here’s what I ended up doing :

`restic forget --dry-run --host myhost --keep-last 30`

Copied the list of snapshots listed a the end _without the brackets_

`restic rewrite --forget --exclude /home/test/foldertoexclude <pastesnapshots>`

Take a coffee, or two

`restic prune`

enjoy !

**Comment by MichaelEischer** [comment]
LGTM. I've made one final change before merging this PR: when just passing`--iexclude-file` the `rewrite` command would complain about missing excludes.

> Everything went very well and all checks are ok (including read-data).

Thanks for testing! Good to hear that the PR indeed worked as expected (it's been in a good shape for some time.

**Review by MichaelEischer — APPROVED** [review]
LGTM2

**Comment by 12nick12** [comment]
Any idea when this will be in the prebuilt binaries?

**Comment by rawtaz** [comment]
It will be in the next release, which is released when it's done.

**Comment by aawsome** [comment]
@12nick12  You might want to have a look at https://beta.restic.net/

**Comment by rawtaz** [comment]
FWIW, we're very close to a new release so stay tuned and you'll have it in your hands soon :)

**Comment by rawtaz** [comment]
@12nick12 It's out now! See: https://restic.net/blog/2023-01-12/restic-0.15.0-released/

**Comment by vvidovic** [comment]
> @12nick12 It's out now! See: https://restic.net/blog/2023-01-12/restic-0.15.0-released/

💚💚💚💚

**Pull request merged**
